### ISSUE TITLE
Memoize module make functions in the embedded scenario

### ISSUE THREAD

**Issue by rehno-lindeque** [elm-lang/elm-compiler#888]
I've been investigating the loading for our elm widgets...

It appears to me that when several widgets are instantiated (embedded), repeated calls to `make` for a large module is very slow. In particular, [Bootstrap.Html](http://package.elm-lang.org/packages/circuithub/elm-bootstrap-html/4.0.0/Bootstrap-Html)`.make()` takes a large amount of time:

![timeline](https://cloud.githubusercontent.com/assets/337811/5917625/a0cacaf2-a627-11e4-86c3-d4ef39928a2a.png)

One way of testing this quickly is as follows:
```javascript
for (var i = 0; i < 100; ++i) {
    Elm.Bootstrap.Html.make({});
}
```
The above takes 12.68s on my laptop.

One quick solution might be to memoize the `make` function. The generated function currently looks like this...
```javascript
Elm.Bootstrap.Html.make = function (_elm) { ... }
```
...I'm not sure if it would be necessary to memoize the function based on the incoming `_elm` RTS, but it seems like that shouldn't be necessary. One edge case is if someone is uses different versions of the same package for different embedded widgets (although, in this case they will have to have different compiled `.js` files for each widget, so these could be independently memoized).

**Comment by rehno-lindeque** [comment]
It looks like there's already some memoization going on in there, it just doesn't work for the embedded scenario:
```javascript
   if (_elm.Bootstrap.Html.values)
   return _elm.Bootstrap.Html.values;
```

**Comment by evancz** [comment]
I talked about this with @jvoigtlaender in [this issue](https://github.com/elm-lang/core/pull/124#issuecomment-70931412). There is a way to share an implementation, but it has tradeoffs. Perhaps we can find a way to share static functions safely?

I think this also will change with @laszlopandy's [dead-code elimination idea](https://github.com/elm-lang/projects#dead-code-elimination).

**Comment by rehno-lindeque** [comment]
Thanks, I'll have a look at that issue.

**Comment by rehno-lindeque** [comment]
With that memoization hack I posted above removes about 5s from the page load time on circuithub. All traces of make is gone:

![timeline2](https://cloud.githubusercontent.com/assets/337811/5920260/f740e53a-a643-11e4-8c05-0c6fa5a6085e.png)

**Comment by evancz** [comment]
Makes sense. If it's possible to figure out how to do it in a safe way, I think it makes sense to do this in code generation. I can't focus on this right now, so it may be a good topic to discuss on elm-dev.

**Comment by evancz** [comment]
I think [this paper](http://www.mpi-sws.org/~rossberg/papers/Rossberg,%20Russo,%20Dreyer%20-%20F-ing%20Applicative%20Functors%20[Abstract].pdf) will go into it more, but the *applicative* vs *generative* distinction seems particularly important. I'm not an expert on this work though.

**Comment by rehno-lindeque** [comment]
Thanks! Don't worry, I know you have to focus :) This hack seems to be good enough temporarily, I'm in a bit of a crunch myself - will probably leave it there for now.

**Comment by evancz** [comment]
I believe this will be resolved in 0.17. Code does not get totally reinitialized anymore.
